src/legacy/server/i18n/index.js

```diff
@@ -20,49 +20,66 @@
 import { resolve } from 'path';
 import globby from 'globby';
 import { i18n, i18nLoader } from '@kbn/i18n';
-
+import { readFile } from 'fs';
 import { fromRoot } from '../../utils';
+import { promisify } from 'util';
+
+const readFileAsync = promisify(readFile);
+
+export async function getTranslationPaths({ cwd, glob }) {
+  const entries = await globby(glob, { cwd });
+
+  const configFiles = await Promise.all(
+    entries
+      .map(entry => resolve(cwd, entry))
+      .map(async configFilePath => ({
+        path: configFilePath,
+        content: await readFileAsync(configFilePath, 'utf8')
+      }))
+  );
+
+  return configFiles
+    .map(({ content, path }) => {
+      try {
+        return JSON.parse(content);
+      } catch (err) {
+        throw new Error(`Failed to parse .i18nrc.json file at ${path}`);
+      }
+    })
+    .reduce((acc, configFile) => acc.concat(...configFile.translations), [])
+    .map(translationPath => resolve(cwd, translationPath));
+}
 
 export async function i18nMixin(kbnServer, server, config) {
   const locale = config.get('i18n.locale');
+  const localeRegExp = new RegExp(`${locale}.json$`);
+  const pluginPaths = config.get('plugins.paths');
 
-  // eslint-disable-next-line max-len
-  const translationsDirs = [fromRoot('src/legacy/ui/translations'), fromRoot('src/legacy/server/translations'), fromRoot('src/core/translations')];
-
-  const groupedEntries = await Promise.all([
-    ...config.get('plugins.scanDirs').map(async path => {
-      const entries = await globby(`*/translations/${locale}.json`, {
-        cwd: path,
-      });
-      return entries.map(entry => resolve(path, entry));
+  const translationPaths = await Promise.all([
+    getTranslationPaths({
+      cwd: fromRoot('.'),
+      glob: `.i18nrc.json`,
     }),
-
-    ...config.get('plugins.paths').map(async path => {
-      const entries = await globby(
-        [`translations/${locale}.json`, `plugins/*/translations/${locale}.json`],
-        {
-          cwd: path,
-        }
-      );
-      return entries.map(entry => resolve(path, entry));
+    ...pluginPaths.map(cwd => getTranslationPaths({ glob: `.i18nrc.json`, cwd })),
+    getTranslationPaths({
+      cwd: fromRoot('plugins'),
+      glob: `*/.i18nrc.json`,
     }),
-
-    ...translationsDirs.map(async path => {
-      const entries = await globby(`${locale}.json`, {
-        cwd: path,
-      });
-      return entries.map(entry => resolve(path, entry));
+    getTranslationPaths({
+      cwd: fromRoot('../kibana-extra'),
+      glob: `*/.i18nrc.json`,
     }),
   ]);
 
-  const translationPaths = [].concat(...groupedEntries);
-  i18nLoader.registerTranslationFiles(translationPaths);
+  const currentTranslationPaths = [].concat(...translationPaths)
```

Comment:
Any reason this couldn't be:

```js
const currentTranslationPaths = translationPaths.filter(translationPath => localeRegExp.test(translationPath));

```

Reply:
translationPaths is an array of resolved promises `Promise.all(...)`, each returns an array. using `concat` flattens the arrays into one.